controllers/useraccount/useraccount_controller.go

```diff
@@ -422,15 +442,43 @@ func (r *Reconciler) deleteIdentity(logger logr.Logger, config membercfg.Configu
 		}
 		return false, nil
 	}
-	logger.Info("deleting the identity resource")
+	logger.Info("deleting the Identity resource")
 	// Delete Identity associated with UserAccount
 	if err := r.Client.Delete(context.TODO(), identity); err != nil {
 		if !errors.IsNotFound(err) {
 			return false, err
 		}
 		return false, nil
 	}
+	logger.Info("deleted the Identity resource")
+	return true, nil
+}
 
+// deleteNSTemplateSet deletes the NSTemplateSet associated with the given UserAccount.
+// Returns bool and error indicating that whether the resource were deleted.
+func (r *Reconciler) deleteNSTemplateSet(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (bool, error) {
+	// Get the NSTemplateSet associated with the UserAccount
+	nstmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := r.Client.Get(context.TODO(),
+		types.NamespacedName{ // same namespace/name as the UserAccount
+			Namespace: userAcc.Namespace,
+			Name:      userAcc.Name},
+		nstmplSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return false, nil
+	}
+	logger.Info("deleting the NSTemplateSet resource")
+	// Delete NSTemplateSet associated with UserAccount
+	if err := r.Client.Delete(context.TODO(), nstmplSet); err != nil {
```

Comment:
You could check if the NSTemplateSet is already being deleted before calling the delete over and over again. It may take some time until it cleans up all namespaces.

Reply:
done in 2056e93